DuckDuckGo/Favicons.swift

```diff
@@ -138,14 +139,15 @@ public class Favicons {
     }
     
     func replaceBookmarksFavicon(forDomain domain: String?, withImage image: UIImage) {
-        
+
         guard let domain = domain,
-            let resource = defaultResource(forDomain: domain),
-            (Constants.bookmarksCache.isCached(forKey: resource.cacheKey) || bookmarksStore.contains(domain: domain)),
-            let options = kfOptions(forDomain: domain, usingCache: .bookmarks) else {
-                loadFavicon(forDomain: domain, intoCache: .bookmarks, fromCache: .tabs)
-                return
-            }
+              let resource = defaultResource(forDomain: domain),
+              let options = kfOptions(forDomain: domain, usingCache: .bookmarks) else { return }
+
+        if !isFaviconCachedForBookmarks(forDomain: domain, resource: resource) {
+            loadFaviconForBookmarks(forDomain: domain)
```

Comment:
I don't think this approach is optimal, as we query all bookmarks on each navigation. I'd try to look for a ways to optimize that, like having some caching layer for bookmarks or leveraging existence of tabs cache (so call above code only in case if we insert into tabs cache - as that won't repeat the calls for subsequent navigations on that domain). There are definitely other ideas to consider as well.

Reply:
I thought about this and tested and I didn't notice any impact to the user experience.  Also I found `BookmarksCachingSearch` which unfortunately is not in Core but when looking at this I saw this constantly makes the same call to `BookmarksCoreDataStorage` when using the app so I presumed that this trade off was acceptable.  Every time the address bar is tapped in the app it makes this CoreData call for all bookmarks.

Assuming this trade off is not acceptable is there an existing data cache structure within the app already (data cache) I could potentially leverage for this or (not)?  I can create an enhancement to the original bookmarks import export so that the domain was separated from the bookmark URL on import and cached to disk in some manner + hopefully leverage something the app already has.

Also if we go this route is this worth discussing if MacOS has this and if that can be shared in BSK?

Reply:
I think the use case for Search and for this is completely different:
 - In case of search: We cache bookmarks and the result of the query is (nearly) always used when user is typing something in autocomplete to present results. There isn't much happening other than that - so it is not affecting user flow but is part of it.
 - In case of favicon we want to apply it when needed - so when we navigate to a domain and obtain info about favicon related to that domain, then we query bookmarks to see if we need to add this favicon to the cache. User does not need that for his current flow and there is plenty of other things we are doing when loading pages, so I'd rather optimize for using less resources and making less actions. 

Much better would be to do that once per domain (host). Best solution would be to Cache bookmarks (actually - just hosts) in memory and invalidate that cache when there are changes to the store (e.g. on bookmarks did save notification).

I wouldn't think about macOS here too deeply as:
- Desktop has much more resources than mobile.
- Memory separation model between apps on the desktop is different than on mobile, so we may not want to keep bookmark URLs in memory all the time there.

Reply:
I've taken a look / spent some time on this and what do you think about moving BookmarksCachingSearch to Core and utilising it's cache for favicons as well?  It seems like this could be shared but you know the app better so happy to defer to you 🙂

Reply:
I think it makes sense. To be on the same page (as I'm out next week and someone will need to pick this up from here):
 - We should be able to instantiate `BookmarksCachingSearch` lazily when needed, then discard it once we receive `dataDidChange` notification from `BookmarksCoreDataStorage`. It should be fine to just keep that in memory as we already do every time user taps URL bar.
 - It probably is also a good idea to change Autosuggest to use the same instance.